cmd/traffic_cache_tool/CacheScan.cc

```diff
@@ -0,0 +1,335 @@
+
+    while (bcount < alt->m_frag_offset_count) {
+      bcount *= 2;
+    }
+    alt->m_frag_offsets =
+      static_cast<uint64_t *>(ats_malloc(bcount * sizeof(uint64_t))); // WRONG - must round up to next power of 2.
+    memcpy(alt->m_frag_offsets, alt->m_integral_frag_offsets, sizeof(alt->m_integral_frag_offsets));
+    memcpy(alt->m_frag_offsets + HTTPCacheAlt::N_INTEGRAL_FRAG_OFFSETS, extra_src, extra);
+    len -= extra;
+  } else if (alt->m_frag_offset_count > 0) {
+    alt->m_frag_offsets = alt->m_integral_frag_offsets;
+  } else {
+    alt->m_frag_offsets = nullptr; // should really already be zero.
+  }
+
+  // request hdrs
+
+  HdrHeap *heap   = (HdrHeap *)(alt->m_request_hdr.m_heap ? (buf + (intptr_t)alt->m_request_hdr.m_heap) : nullptr);
+  HTTPHdrImpl *hh = nullptr;
+  int tmp;
+  if (heap != nullptr) {
+    tmp = this->unmarshal(heap, len, HDR_HEAP_OBJ_HTTP_HEADER, (HdrHeapObjImpl **)&hh, block_ref);
+    if (hh == nullptr || tmp < 0) {
+      ink_assert(!"HTTPInfo::request unmarshal failed");
+      return zret;
+    }
+    len -= tmp;
+    alt->m_request_hdr.m_heap              = heap;
+    alt->m_request_hdr.m_http              = hh;
+    alt->m_request_hdr.m_mime              = hh->m_fields_impl;
+    alt->m_request_hdr.m_url_cached.m_heap = heap;
+  }
+
+  // response hdrs
+
+  heap = (HdrHeap *)(alt->m_response_hdr.m_heap ? (buf + (intptr_t)alt->m_response_hdr.m_heap) : nullptr);
+  if (heap != nullptr) {
+    // tmp = heap->unmarshal(len, HDR_HEAP_OBJ_HTTP_HEADER, (HdrHeapObjImpl **)&hh, block_ref);
+    if (hh == nullptr || tmp < 0) {
+      ink_assert(!"HTTPInfo::response unmarshal failed");
+      return zret;
+    }
+    len -= tmp;
+
+    alt->m_response_hdr.m_heap = heap;
+    alt->m_response_hdr.m_http = hh;
+    alt->m_response_hdr.m_mime = hh->m_fields_impl;
+  }
+
+  alt->m_unmarshal_len = orig_len - len;
+
+  return zret;
+}
+
+Errata
+CacheScan::get_alternates(const char *buf, int length)
+{
+  Errata zret;
+  ink_assert(!(((intptr_t)buf) & 3)); // buf must be aligned
+
+  char *start            = (char *)buf;
+  RefCountObj *block_ref = nullptr;
+
+  while (length - (buf - start) > (int)sizeof(HTTPCacheAlt)) {
+    HTTPCacheAlt *a = (HTTPCacheAlt *)buf;
+
+    if (a->m_magic == CACHE_ALT_MAGIC_MARSHALED) {
+      this->unmarshal((char *)buf, length, block_ref);
+      //        std::cout << "alternate unmarshal failed" << std::endl;
+      //      }
+      auto *url       = a->m_request_hdr.m_http->u.req.m_url_impl;
+      std::string str = std::string(url->m_ptr_scheme, url->m_len_scheme) + "://" + std::string(url->m_ptr_host, url->m_len_host) +
+                        ":" + std::string(url->m_ptr_port, url->m_len_port) + "/" + std::string(url->m_ptr_path, url->m_len_path) +
+                        ";" + std::string(url->m_ptr_params, url->m_len_params) + "?" +
+                        std::string(url->m_ptr_query, url->m_len_query);
+      global_url_map[stripe->hashText].push_back(str);
```

Comment:
If you have multiple threads running, might you not want to lock this before updating?

Reply:
the threads are per span. locking is not really necessary here since the stripes are going to be scanned sequentially in a thread